File: C++/Stack/ValidParentheses.cpp
//Optimal Solution
//Time Complexity: O(n)
//Space Complexity: O(n)
class Solution {
public:
    bool isValid(string s) {
        stack<char> st; // stack
        for (auto it : s) {
            if (it == '(' || it == '[' || it == '{') {
                st.push(it);
            } else {
                if (st.empty()) {
                    return false;
                }
                char el = st.top();
                st.pop();
                if (it != ')' && el == '(' || it != ']' && el == '[' ||
                    it != '}' && el == '{') {
                    return false;
                }
            }
        }
        return st.empty(); // important
    }
};

File: JAVA/Stack/ValidParentheses.java
// Optimal Solution
//Time Complexity: O(n)
//Space Complexity: O(n)
class Solution {
    public boolean isValid(String s) {
        if(s.length()==0){
            return false;
        }
        Stack<Character> st =new Stack<>();
        for(int i=0;i<s.length();i++){
            if(s.charAt(i)=='(' || s.charAt(i)=='{' || s.charAt(i)=='['){
                st.push(s.charAt(i));
            }
            else if(!st.empty()&&((
                st.peek()== '('&& s.charAt(i)==')' || st.peek()== '{'&& s.charAt(i)=='}' || st.peek()== '['&& s.charAt(i)==']')
            )){
                st.pop();
            }
            else{
                return false;
            }
        }
        return st.empty();       
    }
}